requests-for-comments/2022-01-27-future-of-person-model.md

```diff
@@ -0,0 +1,108 @@
+
+
+### Technical implications
+
+### Benefits
+1. Faster queries: remove the need to do JOINs for any insight queries.]
+2. Unlock: Our query results no longer change unless new data has come in - we make it possible to later implement time-series caching to speed up repeated queries over long time frames (>10x improvement expected on repeated calculations).
+4. We can sample by person_id in queries, which will make it possible to build a quick way to speed up queries in huge (1Bn+ events) installations with no additional query time resource cost.
+5. Queries will require less memory and no longer OOM (making it cheaper and more reliable to operate posthog)
+3. We can shard by person_id, simplifying the sharding story a lot and removing the need for cross-shard communication when sharding.
+
+Note that some of these changes have been benchmarked [here](https://github.com/PostHog/posthog/issues/7962)
+
+
+#### Challenges
+
+6. We will need to set up some strong consistency for message ordering in plugin-server and kafka, to make sure messages are processed in the right order per user.
+7. We will store more data in the events table. See below for more details.
+8. Deleting user properties becomes more expensive due to being stored in a redundant way.
+
+<details><summary>Details on cost of storage challenges</summary>
+
+
+Storing person properties in events means we’ll be storing more data.
+
+The exact impact of this depends on the dataset. Some factors about our setup which might not make this into a huge deal:
+
+1. The data is low cardinality and clickhouse is great at compression => we should get really good compression ratios
+2. When [Clickhouse implements semistructured data support](https://github.com/ClickHouse/ClickHouse/pull/23932), compression ratios will get even better.
+
+I’ve done a measurement on a test benchmarking cluster:
+- Original table was 7.56 GiB on disk, 99.13 GiB uncompressed
+- New table (with person_properties/person_id joined in) was 9.16 GiB on disk, 167.88 GiB uncompressed
+- person_properties column was 1.13 GiB on disk, 51.21 GiB uncompressed with a compression ratio of 0.022
+- For reference, `properties` column was 4.49 GiB on disk, 76.28 GiB uncompressed with a compression ratio of 0.059
+
+This measurement was done on a dataset that would compress relatively well (few people), however looking at the uncompressed size, we’d expect at most a 2x disk size increase in most cases.
+
+</details>
+
+## Conclusion
+Scalability of Posthog will be critical for our success, this project lays the foundation for high scalability, removes a critical scaling bottleneck and unblocks future improvements for even larger deployments. There are both technical and product risks associated with this project but they are far outweighed by the long-term benefits.
+
+## Open questions
+
+- How do we handle historical data imports?
+- How do we implement the change on cloud in a cost-effective way?
```

Comment:
Follow-up to our discussion move "We will need to set up some strong consistency for message ordering in plugin-server and kafka, to make sure messages are processed in the right order per user." from challenges to open questions, specifically I would have it here be:
```suggestion
- How do we handle events coming in out of order (both short time differences from network etc and historical data imports)?
- How do we implement the changeover on cloud and self-hosted instances?
```
^ nitted the second one because we might ask feedback from users who are self hosting, this will make clear it's an open question

Reply:
Thank you!